data_structures/linked lists/cll.c

```diff
@@ -0,0 +1,186 @@
+#include<stdio.h>
+#include<stdlib.h>
+
+// Structure definition:
+
+typedef struct NODE{
+	int data;
+	struct NODE* link;
+} node;
+
+// Function Prototypes:
+
+node *InsertAtBeg(node *, int);
+node *InsertAtEnd(node *, int);
+node *InsertAfter(node *, int, int);
+void menu(node *);
+void disp(node *);
+
+// MAIN:
+
+int main()
+{
+    node *last=NULL;
+    menu(last);
+    exit(0);
+}
+
+void menu(node *last)
+{
+    int x,y; char ch;
+    do{
+    disp(last);
+    printf("Enter your Choice:\n(a).Insert at the beginning of the list");
+    printf("\n(b).Insert at the End of the List.\n(c).Insert after a num");
+    printf("ber in the list.\n(e).Exit\n");
+    scanf("\n%c",&ch);
+    switch(ch)
+    {
+	case 'a':printf("Enter the node which you want to insert\t");
+		 scanf("%d",&x);
+		 last = InsertAtBeg(last,x);
+		 break;
+
+	case 'b':printf("Enter the node which you want to insert\t");
+		 scanf("%d",&x);
+		 last = InsertAtEnd(last,x);
+		 break;
+
+	case 'c':printf("Enter the node after which you want to insert\t");
+		 scanf("%d",&y);
+	         printf("Enter the node which you want to insert\t");
+		 scanf("%d",&x);
+		 last = InsertAfter(last,y,x);
+		 break;
+/*
+	case 'd':printf("Enter the node to be deleted\t");
+		 scanf("%d",&y);
+		 last = DeleteNode(last,y);
+		 break;
+*/
+	case 'e':exit(0);break;
+
+	default:printf("Invalid Option.\n");break;
+    }
+    }while(1);
+}
+
+node *InsertAtBeg(node *last, int x)
+{
+    if(!last)
+    {
+    	node *tmp = (node*)malloc(sizeof(node));
+    	tmp->data=x;
+    	tmp->link=tmp;
+	last=tmp;
+	return last;
+    }
+    node *tmp = (node*)malloc(sizeof(node));
+    tmp->data=x;
+    tmp->link=last->link;
+    last->link=tmp;
+    return last;
+}
+
+node *InsertAtEnd(node *last, int x)
+{
+    if(!last){
+	last=InsertAtBeg(last,x);return last;}
+    node *tmp = (node*)malloc(sizeof(node));
+    tmp->data=x;
+    tmp->link=last->link;
+    last->link=tmp;
+    last=tmp;
+    return last;
+}
+
+
+node *InsertAfter(node *last, int y, int x)
+{
+    if(!last)
+	return NULL;
+    node *tmp = (node*)malloc(sizeof(node));
```

Comment:
suppose the value ``y`` is not present in the list, the function should ``free`` this memory.